go/api/base_client.go

```diff
@@ -0,0 +1,151 @@
+// Copyright GLIDE-for-Redis Project Contributors - SPDX Identifier: Apache-2.0
+
+package api
+
+// #cgo LDFLAGS: -L../target/release -lglide_rs
+// #include "../lib.h"
+//
+// void successCallback(void *channelPtr, struct CommandResponse *message);
+// void failureCallback(void *channelPtr, char *errMessage, RequestErrorType errType);
+import "C"
+
+import (
+	"unsafe"
+
+	"github.com/aws/glide-for-redis/go/glide/protobuf"
+	"google.golang.org/protobuf/proto"
+)
+
+// BaseClient defines an interface for methods common to both [RedisClient] and [RedisClusterClient].
+type BaseClient interface {
+	StringCommands
+
+	// Close terminates the client by closing all associated resources.
+	Close()
+}
+
+const OK = "OK"
+
+type payload struct {
+	value *C.struct_CommandResponse
+	error error
+}
+
+//export successCallback
+func successCallback(channelPtr unsafe.Pointer, cResponse *C.struct_CommandResponse) {
+	// TODO: call lib.rs function to free response
+	response := cResponse
+	resultChannel := *(*chan payload)(channelPtr)
+	resultChannel <- payload{value: response, error: nil}
+}
+
+//export failureCallback
+func failureCallback(channelPtr unsafe.Pointer, cErrorMessage *C.char, cErrorType C.RequestErrorType) {
+	// TODO: call lib.rs function to free response
+	resultChannel := *(*chan payload)(channelPtr)
+	resultChannel <- payload{value: nil, error: goError(cErrorType, cErrorMessage)}
+}
+
+type clientConfiguration interface {
+	toProtobuf() *protobuf.ConnectionRequest
+}
+
+type baseClient struct {
+	coreClient unsafe.Pointer
+}
+
+func createClient(config clientConfiguration) (*baseClient, error) {
+	request := config.toProtobuf()
+	msg, err := proto.Marshal(request)
+	if err != nil {
+		return nil, err
+	}
+
+	byteCount := len(msg)
+	requestBytes := C.CBytes(msg)
+	cResponse := (*C.struct_ConnectionResponse)(
+		C.create_client(
+			(*C.uchar)(requestBytes),
+			C.uintptr_t(byteCount),
+			(C.SuccessCallback)(unsafe.Pointer(C.successCallback)),
+			(C.FailureCallback)(unsafe.Pointer(C.failureCallback)),
+		),
+	)
+	defer C.free_connection_response(cResponse)
+
+	cErr := cResponse.connection_error_message
+	if cErr != nil {
+		message := C.GoString(cErr)
+		return nil, &ConnectionError{message}
+	}
+
+	return &baseClient{cResponse.conn_ptr}, nil
+}
+
+// Close terminates the client by closing all associated resources.
+func (client *baseClient) Close() {
+	if client.coreClient == nil {
+		return
+	}
+
+	C.close_client(client.coreClient)
+	client.coreClient = nil
+}
+
+func (client *baseClient) executeCommand(requestType C.RequestType, args []string) (*C.struct_CommandResponse, error) {
+	if client.coreClient == nil {
+		return nil, &ClosingError{"The client is closed."}
+	}
+
+	cArgs := toCStrings(args)
```

Comment:
See my other comment about using `BinarySafeString` instead

Reply:
Replied on the other comment thread.